### ISSUE TITLE
JUnit5Plugin does not find test classes after 'gradle assemble'

### ISSUE THREAD

**Issue by bkimminich** [junit-team/junit5#284]
When I run ```gradlew build``` (Gradle 2.13) locally on Windows 8.1 for my project https://github.com/bkimminich/kata-botwars, I get my tests executed:

```
C:\Data\Github\kata-botwars>gradlew build
:botwars-java:compileJava                                                                                     
:botwars-java:processResources UP-TO-DATE
:botwars-java:classes
:botwars-java:jar
:botwars-java:startScripts                 
:botwars-java:distTar
:botwars-java:distZip
:botwars-java:assemble
:botwars-java:checkstyleMain                                                                     
:botwars-java:compileTestJava                 
:botwars-java:processTestResources UP-TO-DATE
:botwars-java:testClasses
:botwars-java:checkstyleTest UP-TO-DATE
:botwars-java:junit5Test                 
Mai 27, 2016 4:51:07 PM org.junit.gen5.launcher.main.ServiceLoaderTestEngineRegistry loadTestEngines
INFORMATION: Discovered TestEngines with IDs [junit5]
                                         
Test run finished after 100 ms:junit5Test
[        11 tests found     ]
[         1 tests skipped   ]
[        10 tests started   ]
[         0 tests aborted   ]
[        10 tests successful]
[         0 tests failed    ]

:botwars-java:test                                                                      
:botwars-java:check
:botwars-java:build

BUILD SUCCESSFUL

Total time: 9.786 secs
```

Now after pushing and having Travis-CI pick up the commit, the tests are not executed at all:

```
Test run finished after 3 ms
[         0 tests found     ]
[         0 tests skipped   ]
[         0 tests started   ]
[         0 tests aborted   ]
[         0 tests successful]
[         0 tests failed    ]
```
(Full log: https://s3.amazonaws.com/archive.travis-ci.org/jobs/133152764/log.txt)

**Comment by sbrannen** [comment]
I noticed the following in your log:

> java version "1.8.0_31"

AFAIK, JUnit 5 required JDK 1.8 update 60 or higher.

Try changing the JDK version for your build and let us know how that goes.

**Comment by sbrannen** [comment]
It looks like you discovered the solution on your own: https://github.com/bkimminich/kata-botwars/commit/fd3921c5eb07b0be393ecc04a6052c8eb0deb4bb

May we therefore consider this issue resolved?

**Comment by bkimminich** [comment]
Thank you! Adding the following to the ```.travis.yml``` enforced 8u91 but it still didn't execute any tests...

```
addons:
  apt:
    packages:
      - oracle-java8-installer
```

**Comment by sbrannen** [comment]
I'll check it out locally

**Comment by sbrannen** [comment]
By the way, you need to set the file permissions for `gradlew` appropriately. After doing so locally, I see this with `git diff`:

```
[kata-botwars (master)]$ git diff
diff --git a/gradlew b/gradlew
old mode 100644
new mode 100755
diff --git a/gradlew.bat b/gradlew.bat
old mode 100644
new mode 100755
```

You should execute `./gradlew wrapper`, commit changes, and push to `master`.

**Comment by sbrannen** [comment]
I just ran your Gradle build with JDK `1.8.0_92`, and I see this:

```
INFO: Discovered TestEngines with IDs [junit5]

Test run finished after 3 ms
[         0 tests found     ]
[         0 tests skipped   ]
[         0 tests started   ]
[         0 tests aborted   ]
[         0 tests successful]
[         0 tests failed    ]
```

**Comment by marcphilipp** [comment]
`master` doesn't actually contain any tests...

**Comment by sbrannen** [comment]
@marcphilipp, exactly!

Just what I was about to say....

@bkimminich, you apparently forgot to check in your `src/test/java` folder for the `botwars-java` project.

That would explain why the tests are executed locally but not on the server... because they don't exist on the server. 😉

**Comment by marcphilipp** [comment]
I think @bkimminich is referring to the `develop` branch, right?

**Comment by marcphilipp** [comment]
https://travis-ci.org/bkimminich/kata-botwars/builds/133438724

**Comment by bkimminich** [comment]
I'm worried about my develop branch, not master. There I also get no tests run... But I'll check, if I seriously never pushed those there... 😬

Am 27. Mai 2016 19:45:46 MESZ, schrieb Sam Brannen <notifications@github.com>:

**Comment by sbrannen** [comment]
The problem is has to do with the order in which you execute Gradle commands on the CI server.

I haven't found the exact root problem, but it appears to be an issue with the gradle commands you supply for the coverity scan.

As on the CI server, the following does **not** run the tests locally for me:

```
./gradlew assemble
./gradlew check
```

But the following **does run** the tests locally for me:

```
./gradlew clean assemble check
```

So it looks like the `clean` task is not actually cleaning up everything if you execute `assemble` separately.

In other words, I don't think there is an issue with JUnit 5 but rather with your Gradle task configuration.

**Comment by bkimminich** [comment]
That addon for CoverityScan will only be run on the ```coverity_scan``` branch. From the log I see that ```gradle assemble``` and ```gradle check``` were both executed.

To test if the build artifacts are actually there I'll include a CheckStyle violation. If it's a general issue post-```assemble``` it shouldn't be found by CheckStyle either.

**Comment by marcphilipp** [comment]
I don't know why yet, but `project.sourceSets.test.runtimeClasspath.files` only seems to contain one directory, namely `/Users/marcp/Desktop/kata-botwars/botwars-java/build/classes/main`. Thus, test classes are never scanned.

**Comment by sbrannen** [comment]
What I can observe is that....

Running `clean assemble` does not compile the test sources.

And then running `clean check` will in fact compile the test sources but seemingly *after* the `junit5Test` task has been executed. Thus there physically are no tests when the JUnit 5 Gradle plugin executes.

**Comment by sbrannen** [comment]
Oddly enough, the following does run the tests... but only for the second execution of `clean check`... which seems a bit like a bug in Gradle or otherwise a bug in the JUnit 5 Gradle plugin.

```
gradlew clean assemble
gradlew clean check
gradlew clean check
```

**Comment by bkimminich** [comment]
Yes, that CheckStyle error I put in one of the test data builder (thus in ```src/test/...```) was found by CheckStyle's test-phase scan. As that happens *before* JUnit's task, the sources *are there* at least for a moment... :smile:

```
:botwars-java:assemble
BUILD SUCCESSFUL
Total time: 28.509 secs
18.78s$ ./gradlew check
:botwars-java:compileJava UP-TO-DATE
:botwars-java:processResources UP-TO-DATE
:botwars-java:classes UP-TO-DATE
:botwars-java:checkstyleMain
Download https://jcenter.bintray.com/com/puppycrawl/tools/checkstyle/5.9/checkstyle-5.9.pom
Download https://jcenter.bintray.com/antlr/antlr/2.7.7/antlr-2.7.7.pom
Download https://jcenter.bintray.com/commons-beanutils/commons-beanutils-core/1.8.3/commons-beanutils-core-1.8.3.pom
Download https://jcenter.bintray.com/commons-cli/commons-cli/1.2/commons-cli-1.2.pom
Download https://jcenter.bintray.com/com/google/guava/guava-jdk5/14.0.1/guava-jdk5-14.0.1.pom
Download https://jcenter.bintray.com/com/google/guava/guava-parent-jdk5/14.0.1/guava-parent-jdk5-14.0.1.pom
Download https://jcenter.bintray.com/com/puppycrawl/tools/checkstyle/5.9/checkstyle-5.9.jar
Download https://jcenter.bintray.com/antlr/antlr/2.7.7/antlr-2.7.7.jar
Download https://jcenter.bintray.com/commons-beanutils/commons-beanutils-core/1.8.3/commons-beanutils-core-1.8.3.jar
Download https://jcenter.bintray.com/commons-cli/commons-cli/1.2/commons-cli-1.2.jar
Download https://jcenter.bintray.com/com/google/guava/guava-jdk5/14.0.1/guava-jdk5-14.0.1.jar
:botwars-java:compileTestJava
:botwars-java:processTestResources UP-TO-DATE
:botwars-java:testClasses
:botwars-java:checkstyleTest[ant:checkstyle] botwars-java/src/test/java/de/kimminich/kata/botwars/BotBuilder.java:5: Line is longer than 120 characters (found 243).
 FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':botwars-java:checkstyleTest'.
> Checkstyle rule violations were found. See the report at: file:///home/travis/build/bkimminich/kata-botwars/botwars-java/build/reports/checkstyle/test.html
```

https://travis-ci.org/bkimminich/kata-botwars/builds/133453588

**Comment by marcphilipp** [comment]
I think `project.sourceSets.test.runtimeClasspath`gets evaluated too early (before compiling test classes) and thus the test output folder does not exist at that time.

**Comment by sbrannen** [comment]
The bizarre thing is that you only hit this problem if you execute `assemble` in a _separate_ command before running `check` in a subsequent command.

If you execute `clean` and then `check` in two separate commands, it works.

If you execute `assemble` and then `clean` and then `check` in three separate commands, it works.

If you execute `clean assemble check` in the same command, it works.

**Comment by sbrannen** [comment]
@aalmiray, have you ever encountered behavior like this with Gradle?

**Comment by marcphilipp** [comment]
I think I've figured it out:

When you call `assemble` on a clean working directory, Gradle executes `compileJava` but not `compileTestJava`. Thus, when you run `check` in the next command, the `build/classes/main` exists while `build/classes/test` does not. Thus, the current (and the ALPHA) implementation of the `JUnit5Plugin` only adds `build/classes/main` to the arguments when calling `ConsoleRunner` (see [source](https://github.com/junit-team/junit5/blob/r5.0.0-ALPHA/junit-gradle/src/main/groovy/org/junit/gen5/gradle/JUnit5Plugin.groovy#L94-L97)). Thus, only `build/classes/main` will be scanned for test classes. -> No tests

When you call `clean assemble check` in the same command neither `build/classes/main` nor `build/classes/test` exist. Thus, the current (and the ALPHA) implementation of the `JUnit5Plugin` does not add any directory as an extra argument when calling `ConsoleRunner` (see [source](https://github.com/junit-team/junit5/blob/r5.0.0-ALPHA/junit-gradle/src/main/groovy/org/junit/gen5/gradle/JUnit5Plugin.groovy#L94-L97)). `ConsoleRunner` will then look for directories in the system property `java.class.path`. Since test sources have been compiled by `compileTestJava` in the meantime, it will find them.

The crucial point seems to be that `JUnit5Plugin` checks directories in `project.sourceSets.test.runtimeClasspath` at the very beginning of the build when creating the `junit5Test` task.

**Comment by aalmiray** [comment]
I'm afraid I have not seen this problem, but based on the previous description it may be related to the input/output configuration of the JUnit5 test task created by the `Junit5Plugin`.

**Comment by sbrannen** [comment]
Thanks for the tip, @aalmiray.

We will continue to investigate.

**Comment by marcphilipp** [comment]
@bkimminich I've fixed this issue in cc1287b and submitted a pull request to your repo to use the latest snapshot (see bkimminich/kata-botwars#2). We are going to release 5.0.0-M1 next week so you can change it to that version then. Please verify that it now works on Travis and let us know.

**Comment by marcphilipp** [comment]
I'm closing this issue because it now works on Travis https://travis-ci.org/bkimminich/kata-botwars/builds/133487461.

**Comment by bkimminich** [comment]
Thanks!!!
